sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsublite/internal/UnboundedReaderImpl.java

```diff
@@ -0,0 +1,136 @@
+
+import com.google.api.core.ApiService.State;
+import com.google.cloud.pubsublite.Offset;
+import com.google.cloud.pubsublite.proto.SequencedMessage;
+import com.google.protobuf.util.Timestamps;
+import java.io.IOException;
+import java.util.NoSuchElementException;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import org.apache.beam.sdk.io.UnboundedSource;
+import org.apache.beam.sdk.io.UnboundedSource.UnboundedReader;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.joda.time.Instant;
+
+public class UnboundedReaderImpl extends UnboundedReader<SequencedMessage> {
+
+  private final UnboundedSource<SequencedMessage, CheckpointMarkImpl> source;
+  private final MemoryBufferedSubscriber subscriber;
+  private final TopicBacklogReader backlogReader;
+  private final BlockingCommitter committer;
+
+  private Offset fetchOffset;
+  private Optional<Instant> lastMessageTimestamp = Optional.empty();
+  private boolean advanced = false;
+
+  UnboundedReaderImpl(
+      UnboundedSource<SequencedMessage, CheckpointMarkImpl> source,
+      MemoryBufferedSubscriber subscriber,
+      TopicBacklogReader backlogReader,
+      BlockingCommitter committer,
+      Offset initialOffset) {
+    checkArgument(initialOffset.equals(subscriber.fetchOffset()));
+    this.source = source;
+    this.subscriber = subscriber;
+    this.backlogReader = backlogReader;
+    this.committer = committer;
+    this.fetchOffset = initialOffset;
+  }
+
+  @Override
+  public SequencedMessage getCurrent() throws NoSuchElementException {
+    if (!advanced) {
+      throw new NoSuchElementException();
+    }
+    return subscriber.peek().get();
+  }
+
+  @Override
+  public Instant getCurrentTimestamp() throws NoSuchElementException {
+    return getTimestamp(getCurrent());
+  }
+
+  private static Instant getTimestamp(SequencedMessage message) {
+    return Instant.ofEpochMilli(Timestamps.toMillis(message.getPublishTime()));
+  }
+
+  @Override
+  public void close() throws IOException {
+    try (AutoCloseable c1 = backlogReader;
+        AutoCloseable c2 = committer;
+        AutoCloseable c3 = asCloseable(subscriber)) {
+    } catch (Exception e) {
+      throw new IOException(e);
+    }
+  }
+
+  @Override
+  public boolean start() throws IOException {
+    try {
+      subscriber.startAsync().awaitRunning(1, TimeUnit.MINUTES);
+    } catch (Exception e) {
+      throw new IOException(e);
+    }
+    return advance();
+  }
+
+  @Override
+  public boolean advance() throws IOException {
+    if (!subscriber.state().equals(State.RUNNING)) {
+      throw new IOException("Subscriber failed: ", subscriber.failureCause());
+    }
+    if (advanced) {
+      SequencedMessage peeked = subscriber.peek().get();
+      Offset nextOffset = Offset.of(peeked.getCursor().getOffset() + 1);
+      checkState(nextOffset.value() > fetchOffset.value());
+      fetchOffset = nextOffset;
+      lastMessageTimestamp = Optional.of(getTimestamp(peeked));
```

Comment:
Seems like lastMessageTimestamp and fetchOffset will not be set for the first record ?

Reply:
It depends on what the watermark and checkpointMark should reflect- whether they should reflect the current message or all previous messages. If it is all previous messages, the behavior here should be correct: the watermark and checkpoint mark reflect all messages processed in the past. If it should reflect the current message as well, this behavior should be updated to return getCurrentTimestamp for getWatermark and getCurrent().offset() + 1 for the checkpointMark

Reply:
Fixed to match documented semantics in UnboundedReader interface: `An element is considered 'read' if it could be returned by a call to {@link #getCurrent}`